iroh-gossip/src/proto/state.rs

```diff
@@ -0,0 +1,253 @@
+    config: Config,
+    rng: R,
+    states: HashMap<TopicId, topic::State<PA, R>>,
+    outbox: Outbox<PA>,
+    conns: ConnsMap<PA>,
+}
+
+impl<PA: PeerAddress, R: Rng + Clone> State<PA, R> {
+    pub fn new(me: PA, me_data: PeerData, config: Config, rng: R) -> Self {
+        Self {
+            me,
+            me_data,
+            config,
+            rng,
+            states: Default::default(),
+            outbox: Default::default(),
+            conns: Default::default(),
+        }
+    }
+
+    pub fn endpoint(&self) -> &PA {
+        &self.me
+    }
+
+    pub fn state(&self, topic: &TopicId) -> Option<&topic::State<PA, R>> {
+        self.states.get(topic)
+    }
+
+    pub fn state_mut(&mut self, topic: &TopicId) -> Option<&mut topic::State<PA, R>> {
+        self.states.get_mut(topic)
+    }
+
+    pub fn has_active_peers(&self, topic: &TopicId) -> bool {
+        self.state(topic)
+            .map(|s| s.has_active_peers())
+            .unwrap_or(false)
+    }
+
+    pub fn handle(
+        &mut self,
+        event: InEvent<PA>,
+        now: Instant,
+    ) -> impl Iterator<Item = OutEvent<PA>> + '_ {
+        let event: InEventMapped<PA> = event.into();
+
+        // todo: add command to leave a topic
+        match event {
+            InEventMapped::TopicEvent(topic, event) => {
+                // when receiving messages, update our conn map to take note that this topic state may want
+                // to keep this connection
+                // TODO: this is a lot of hashmap lookups, maybe there's ways to optimize?
+                if let topic::InEvent::RecvMessage(from, _message) = &event {
+                    self.conns.entry(*from).or_default().insert(topic);
+                }
+                // when receiving a join command, initialize state if it doesn't exist
+                if let topic::InEvent::Command(topic::Command::Join(_peer)) = &event {
+                    if let hash_map::Entry::Vacant(e) = self.states.entry(topic) {
+                        e.insert(topic::State::with_rng(
+                            self.me,
+                            self.me_data.clone(),
+                            self.config.clone(),
+                            self.rng.clone(),
+                        ));
+                    }
+                }
+                if let hash_map::Entry::Vacant(e) = self.states.entry(topic) {
+                    e.insert(topic::State::with_rng(
+                        self.me,
+                        self.me_data.clone(),
+                        self.config.clone(),
+                        self.rng.clone(),
+                    ));
+                }
+
+                // pass the event to the state handler
+                if let Some(state) = self.states.get_mut(&topic) {
+                    let out = state.handle(event, now);
+                    for event in out {
+                        handle_out_event(topic, event, &mut self.conns, &mut self.outbox);
```

Comment:
Can any of these block/take longer? if so should this be parallelized potentially?

Reply:
Currently I don't think that's easily possible, because they need the mutable references. Also, with current settings the numbers will be quite low usually.
There might be places where we could indeed optimize by parallelizing things though. I'll think a bit more if/where this is feasable. Also, maybe we should run some benchmarks first if its actually worth it?

Reply:
totally, was more thinking out loud/wondering